File: AU-Serv/Module/mFileFunction.vb
Option Strict On
Option Explicit On

Module mFileFunction

    Public Const PATHBACKSLASH As Short = 7
    Public Const NODELETE As Short = 76
    Public Const BACKSLASH As Short = 6
    Public Const SLASH As Short = 5
    Public Const PATH As Short = 4
    Public Const FILENAME As Short = 3
    Public Const EXTENSION_ONLY As Short = 2
    Public Const FILENAME_ONLY As Short = 1
    Public Const PATH_FILENAME_ONLY As Short = 8

    'Private MaClasse As clsProcess

    Public Function ParsePath(ByVal szPath As String, ByVal nOperation As Short) As String

        Dim nPosSlash, nPosPoint, nLenPath As Integer

        nPosPoint = InStrRev(szPath, ".") - 1
        nPosSlash = InStrRev(szPath, "\")
        nLenPath = Len(szPath)

        If nOperation = EXTENSION_ONLY Then
            Return Right(szPath, nLenPath - nPosPoint)
        ElseIf nOperation = FILENAME_ONLY Then
            Return Mid(szPath, nPosSlash + 1, nPosPoint - nPosSlash)
        ElseIf nOperation = FILENAME Then
            Return Right(szPath, nLenPath - nPosSlash)
        ElseIf nOperation = PATH Then
            Return Left(szPath, nPosSlash)
        ElseIf nOperation = PATH_FILENAME_ONLY Then
            Return Left(szPath, nPosPoint)
        Else
            Return szPath
        End If


    End Function

    Public Function ReadInfoToFile(ByVal NomFic As String, Optional ByRef nOperation As Short = 0) As String

        Dim svInfo As String
        'Dim FileNumber As Integer

        svInfo = ""

        If IO.File.Exists(NomFic) Then

            svInfo = IO.File.ReadAllText(NomFic, System.Text.Encoding.Default)

            Dim svInfoSplit() As String = Split(svInfo, vbCrLf)

            For u As Integer = svInfoSplit.Length - 1 To 0 Step -1
                If svInfoSplit(u) <> "" Then
                    ReDim Preserve svInfoSplit(u)
                    Exit For
                End If
            Next

            svInfo = Join(svInfoSplit, vbCrLf)


            'FileNumber = FreeFile()
            'FileOpen(FileNumber, NomFic, OpenMode.Input)
            'svInfo = LineInput(FileNumber)
            'FileClose(FileNumber)
            'svInfo = Trim(svInfo)

            If nOperation <> NODELETE Then
                IO.File.Delete(NomFic)
            End If
        End If

        ReadInfoToFile = svInfo

    End Function

    Public Sub WriteInfoToFile(ByVal NomFic As String, ByVal svInfo As String)


        If IO.File.Exists(NomFic) Then
            IO.File.Delete(NomFic)
        End If

        IO.File.WriteAllText(NomFic, svInfo, System.Text.Encoding.Default)

        'Dim FileNumber As Integer
        'FileNumber = FreeFile()
        'FileOpen(FileNumber, NomFic, OpenMode.Output)
        ''WriteLine(FileNumber, svInfo)
        'Print(FileNumber, svInfo)
        'FileClose(FileNumber)

    End Sub


    Public Function Add(ByVal svString As String, ByVal nOperation As Short, Optional ByVal svPath As String = "") As String

        If nOperation = SLASH Then
            If (svString <> "") And (svString <> "/") Then
                If (Right(svString, 1) <> "/") Then svString = svString & "/"
                If (Left(svString, 1) <> "/") Then svString = "/" & svString
            Else
                svString = "/"
            End If

        ElseIf nOperation = BACKSLASH Then
            If (svString <> "") Then
                If Right(svString, 1) <> "\" Then svString = svString & "\"
            End If

        ElseIf nOperation = PATH Then
            If (svPath <> "") Then
                svPath = Add(svPath, BACKSLASH)
                If Mid(svString, 2, 1) <> ":" And Mid(svString, 2, 1) <> "\" Then svString = svPath & svString
            End If

        ElseIf nOperation = PATHBACKSLASH Then
            svString = Add(svString, BACKSLASH)
            svString = Add(svString, PATH, svPath)

        End If

        Return svString

    End Function

    Public Function IsDirectoryWritable(ByVal path As String) As Boolean
        Dim info As New IO.DirectoryInfo(path)
        Return (info.Attributes And IO.FileAttributes.ReadOnly) <> IO.FileAttributes.ReadOnly
    End Function

    Public Function IsFileReadOnly(ByVal sFile As String) As Boolean
        Dim objFileInfo As IO.FileInfo = My.Computer.FileSystem.GetFileInfo(sFile)
        Return objFileInfo.IsReadOnly
    End Function

End Module

File: MultiUserEDI/MultiUserEDI/mFileFunction.cs
// MultiUserEDI.mFileFunction
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;
using System.IO;
using System.Text;
using MultiUserEDI.My;

namespace MultiUserEDI
{
    [StandardModule]
    internal sealed class mFileFunction
    {
        public const short PATHBACKSLASH = 7;

        public const short NODELETE = 76;

        public const short BACKSLASH = 6;

        public const short SLASH = 5;

        public const short PATH = 4;

        public const short FILENAME = 3;

        public const short EXTENSION_ONLY = 2;

        public const short FILENAME_ONLY = 1;

        public const short PATH_FILENAME_ONLY = 8;

        public static string ParsePath(string szPath, short nOperation)
        {
            //return System.IO.Path.GetDirectoryName(szPath);

            //checked
            //{
            int num = Strings.InStrRev(szPath, ".") - 1;
            int num2 = Strings.InStrRev(szPath, "\\");
            int num3 = Strings.Len(szPath);
            //return nOperation switch

            //    {

            //        2 => Strings.Right(szPath, num3 - num),
            //        1 => Strings.Mid(szPath, num2 + 1, num - num2),
            //        3 => Strings.Right(szPath, num3 - num2),
            //        4 => Strings.Left(szPath, num2),
            //        8 => Strings.Left(szPath, num),
            //        _ => szPath,
            //    };

            return Strings.Right(szPath, num3 - num) + Strings.Mid(szPath, num2 + 1, num - num2) +
                   Strings.Right(szPath, num3 - num2) + Strings.Left(szPath, num2) + Strings.Left(szPath, num) + szPath;

        }

        public static string ReadInfoToFile(string NomFic, short nOperation = 0)
        {
            string result = "";
            checked
            {
                if (File.Exists(NomFic))
                {
                    result = File.ReadAllText(NomFic, Encoding.Default);
                    string[] array = Strings.Split(result, "\r\n");
                    for (int i = array.Length - 1; i >= 0; i += -1)
                    {
                        if (Operators.CompareString(array[i], "", TextCompare: false) != 0)
                        {
                            array = (string[])Utils.CopyArray(array, new string[i + 1]);
                            break;
                        }
                    }
                    result = Strings.Join(array, "\r\n");
                    if (nOperation != 76)
                    {
                        File.Delete(NomFic);
                    }
                }
                return result;
            }
        }

        public static void WriteInfoToFile(string NomFic, string svInfo)
        {
            if (File.Exists(NomFic))
            {
                File.Delete(NomFic);
            }
            File.WriteAllText(NomFic, svInfo, Encoding.Default);
        }

        public static string Add(string svString, short nOperation, string svPath = "")
        {
            switch (nOperation)
            {
                case 5:
                    if ((Operators.CompareString(svString, "", TextCompare: false) != 0) & (Operators.CompareString(svString, "/", TextCompare: false) != 0))
                    {
                        if (Operators.CompareString(Strings.Right(svString, 1), "/", TextCompare: false) != 0)
                        {
                            svString += "/";
                        }
                        if (Operators.CompareString(Strings.Left(svString, 1), "/", TextCompare: false) != 0)
                        {
                            svString = "/" + svString;
                        }
                    }
                    else
                    {
                        svString = "/";
                    }
                    break;
                case 6:
                    if (Operators.CompareString(svString, "", TextCompare: false) != 0 && Operators.CompareString(Strings.Right(svString, 1), "\\", TextCompare: false) != 0)
                    {
                        svString += "\\";
                    }
                    break;
                case 4:
                    if (Operators.CompareString(svPath, "", TextCompare: false) != 0)
                    {
                        svPath = Add(svPath, 6);
                        if ((Operators.CompareString(Strings.Mid(svString, 2, 1), ":", TextCompare: false) != 0) & (Operators.CompareString(Strings.Mid(svString, 2, 1), "\\", TextCompare: false) != 0))
                        {
                            svString = svPath + svString;
                        }
                    }
                    break;
                case 7:
                    svString = Add(svString, 6);
                    svString = Add(svString, 4, svPath);
                    break;
            }
            return svString;
        }

        public static bool IsDirectoryWritable(string path)
        {
            DirectoryInfo directoryInfo = new DirectoryInfo(path);
            if (!directoryInfo.Exists)
            {
                return false;
            }
            return (directoryInfo.Attributes & FileAttributes.ReadOnly) != FileAttributes.ReadOnly;
        }

        public static bool IsFileReadOnly(string sFile)
        {
            return MyProject.Computer.FileSystem.GetFileInfo(sFile).IsReadOnly;
        }
    }
}
